### DISCUSSION TITLE
DNC Custom Combos

### DISCUSSION THREAD

**Discussion by drarko-monn** [PrincessRTFM/XIVComboPlugin#79] [Show and tell]
Hi

Some time ago I created some DNC Macros, very close to the logic implemented by this plugins, but in some cases going one step forward and chaining Single Target Combo with Fan Dance I-III-IV and Saber Dance if available, and AoE Combo with Fan Dance II-III-IV plus Saber Dance too

Of course Macros are really slow and buggy but in some cases I still use it.

I wonder if you could provide some guidance, if it is possible to implement something similar, I have some devs skills and can provide a PR, but I dont know where start looking. Thanks in advance

This are the Macros

```
/ac  "Fan Dance IV" <t>
/ac "Starfall Dance" <t>
/ac "Devilment"
/ac "Saber Dance" <t>
/ac "Fan Dance III" <t>
/ac "Fan Dance" <t>
/ac "Flourish"
/ac "Reverse Cascade" <t>
/ac "Cascade" <t>
/macroicon Cascade
```

```
/ac  "Fan Dance IV" <t>
/ac "Starfall Dance" <t>
/ac "Devilment"
/ac "Saber Dance" <t>
/ac "Fan Dance III" <t>
/ac "Fan Dance" <t>
/ac "Flourish"
/ac "Fountainfall" <t>
/ac "Fountain" <t>
/macroicon Fountain
```

```
/targetenemy
/ac  "Fan Dance IV" <t>
/ac "Starfall Dance" <t>
/ac "Devilment"
/ac "Saber Dance" <t>
/ac "Fan Dance III" <t>
/ac "Fan Dance II"
/ac "Flourish"
/ac "Rising Windmill" 
/ac "Windmill"
/macroicon Windmill
```

```
/targetenemy
/ac  "Fan Dance IV" <t>
/ac "Starfall Dance" <t>
/ac "Devilment"
/ac "Saber Dance" <t>
/ac "Fan Dance III" <t>
/ac "Fan Dance II"
/ac "Flourish"
/ac "Rising Windmill" 
/ac "Windmill"
/macroicon Windmill
```

**Comment by VariableVixen** [comment]
I can add Flourish to the smart weave options, but Saber Dance and Starfall Dance are both weaponskills, not abilities, so they're on GCD and cannot be woven. Devilment is an ability, but I'm a little on the fence about adding that to the weaves, because it unlocks Starfall for a limited duration.

-----

On the one hand, I don't think there are many cases where you _wouldn't_ want to use Devilment and Starfall basically on CD - if a trash pack is about to be wiped out and you want to save it for the next set/boss, or if the boss is about to invuln and everyone should hold their burst windows, and that's about all I can think of.

On the other hand, those situations _definitely_ come up.

On the first tail, Starfall is only usable after Devilment, and that would be a _weave_, so you could just wait out the GCD and _not_ weave it in.

On the second tail, I'd need to arrange the priorities for what to weave first when multiple options are available, and Devilment should be used as soon as you can... except for one of those situations where you _don't_ want to use it immediately.

...on the third tail, I _could_ arrange it such that the Fan Dances would weave in first when they have a configurable number of charges (1/2) or the effect is below a certain time left (3/4) in order to save them, and otherwise be Devilment first.

And on the fourth tail... if you're in one of those cases where you _do_ want to hold Devilment, is saving up your feathers a _bad_ idea? You'd be holding them for the burst window. So maybe the time-limited dances should weave first, then Devilment, then the charge-based dances.

-----

I've actually already been thinking about adding an Espirit Saver feature that replaces attacks with Saber Dance when your gauge is at least a configurable threshold, I just haven't gotten around to it yet. If the range on that option is `[50, 100]` then you could set it to 50 to use as soon as SD is up, or 100 to only use when you're _at_ cap.

-----

I'll add smart Flourish weaving where it won't weave in if you already have one of the effects it grants, and I'll add the gauge saver, and I might look at reordering the checks for the Fan Dance weaves, but let me know your thoughts on the whole Devilment weaving bit in the middle chunk up there. I'm not against the idea, I'm just not sure about the priority order for what to weave _first_ when multiple options are up at the same time.

**Reply by VariableVixen** [reply]
Update: the Flourish weave and SD gauge saver/spender options are implemented locally, I want to test them a little bit first just to be sure everything's working as intended. All that's left is the Devilment and Starfall options, which I'm still thinking over.

**Reply by VariableVixen** [reply]
Second update: I've added a Starfall Saver feature to turn the multibuttons into Starfall Dance when it's available but the time left is no more than a set amount. All that's left on your request list (as I understand it) is the weaving of Devilment. Still not sure of where to slot it into the priorities. Leaning towards _after_ time-gated Fan Dances, but _before_ charge-based ones and Flourish, but lemme know your thoughts on it when you get a chance. I guess I could also add a suboption to move it to the front of the list...

**Comment by VariableVixen** [comment]
As for where to look if you want to write your own combos, you'd want to go to `Combos/DNC.cs` for the replacers, `CustomComboPreset.cs` for the options in the settings window, and `Config/PluginConfiguration.cs` for the configurable detail settings. Probably also useful are `CustomCombo.cs` for the base class inherited by all replacers, which includes a number of useful shortcuts and methods for checking all sorts of conditions.

Obviously, if you want to write for a job other than Dancer, you'd be looking in a different `Combos/*.cs` file, but they're all named by job abbreviation, with `Combos/Common.cs` being used for things that show up in multiple jobs, and `Combos/DOH.cs` and `Combos/DOL.cs` being for Disciple of the Hand and Land (respectively). At present, DOH jobs haven't got any combos, and I think they don't even use the hooked functions so I can't _add_ any, but it's there for future expansion at least.
